### PR TITLE
Bug fix: truncating overlapping blocks may require re-sorting of the

### PR DESCRIPTION
**Pull request by nh13**
blocks.

The blocks are sorted by start position then length.  If we update a block's start position, it cannot be assumed to have the smallest start position in the list of remaining blocks, so we need to re-insert it and re-sort the list of remaining blocks.  This is in the part of the code that truncates overlapping phased blocks.

### CODE REVIEW

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala [hunk 1]
```diff
@@ -404,49 +405,65 @@ object PhaseBlock extends LazyLogging {
       progress.record(ctx.getContig, ctx.getStart)
     }
 
+    val blockSet = new util.TreeSet[PhaseBlock](new Comparator[PhaseBlock] {
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: Scala has a mutable TreeSet in `scala.collecitons.mutable`, why use the `java.util` one instead?

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala [hunk 2]
```diff
@@ -404,49 +405,65 @@ object PhaseBlock extends LazyLogging {
...
+      def compare(a: PhaseBlock, b: PhaseBlock): Int = {
+        if (a.start < b.start) -1
+        else if (a.start == b.start) b.length - a.length
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: Could you comment that this is to return the shorter block first (I think)?

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala [hunk 3]
```diff
@@ -404,49 +405,65 @@ object PhaseBlock extends LazyLogging {
...
-          else {
-            blocks.append(left)
-            left = right
-          }
+
+    var left = blockSet.pollFirst()
+    while (!blockSet.isEmpty) {
+      val right = blockSet.pollFirst()
+
+      if (right.getStart < left.getStart) {
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: Checks like this can be written as just:
`require(left.getStart <= right.getStart, s"left: $left right: $right")`

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala [hunk 4]
```diff
@@ -404,49 +405,65 @@ object PhaseBlock extends LazyLogging {
...
+        throw new IllegalArgumentException(s"left: $left right: $right")
+      }
+      require(left.getStart <= right.getStart)
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: Oh, I see you know that.  Why not just put the message in here and scrap the above?

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala [hunk 5]
```diff
@@ -404,49 +405,65 @@ object PhaseBlock extends LazyLogging {
...
+      if (left.overlaps(right)) {
+        if (!modifyBlocks) {
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: Could also be a `require(modifyBlocks, "....")`

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala [hunk 6]
```diff
@@ -404,49 +405,65 @@ object PhaseBlock extends LazyLogging {
...
+          throw new IllegalStateException(s"Block $left overlaps $right")
+        }
+        if (left.encloses(right)) { // keep the left block only
+          logger.info(s"Removing $right enclosed in $left")
+          // do nothing
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: `// do nothing (keep left == left) because left may overlap subsequent blocks too!`

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala [hunk 7]
```diff
@@ -404,49 +405,65 @@ object PhaseBlock extends LazyLogging {
...
+          left = right
         }
-        // don't forget the last one
+        else { // truncate right
+          logger.info(s"Truncating $right which overlaps $left");
+          blocks.append(left)
+          // Since the start position of the right block is changed, and we are going to use it in the next iteration,
+          // then re-insert it into the set and poll.
+          blockSet.add(right.copy(start=left.getEnd+1))
+          left = blockSet.pollFirst()
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: I think even if you truncate right, you should assign `left = right` here instead of `left = blockSet.pollFirst()`, since presumably `right` could also overlap with the subsequent block, but you'll never do that comparison with the code as is.

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala (hunk 1 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: Since it has `pollFirst`, which retrieves the first element and removes it all in one shot!

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala (hunk 2 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: Done

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala (hunk 3 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: Done

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala (hunk 4 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: Done.

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala (hunk 5 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: Done.

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala (hunk 7 above)
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: This is the part that was buggy.  We require that `left.start <= right.start` at the start of the loop. If we assign `left = right`, then in the requirement many not hold.  An example, which I will add to a test, would be: `(1, 10), (8, 12), (9, 13)`.  If we do as you suggested, then in the first iteration we would truncate the right to `(11, 12)` and assign `left = (11,12)`, and in the next loop we would assign `right = (9, 13)` (`right = blockSet.pollFirst()`), and now our requirement that `left.start <= right.start` at the start of the loop fails.  What we do instead is re-insert `right = (11, 12)` into the tree-set, then call `pollFirst` which assigns `left = (9, 13)`, and so in the next loop, `right = (11, 12)` and the requirement holds.  This is all to do with sorting by start position.

**Comment by nh13** [comment]
@tfenne I am going to leave the `if/else` as is.  I added two tests that should cover out the corner case I was trying to solve.

File: src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala (hunk 7 above)
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: I stand by my earlier comment that I find this loop super-confusing 😛

Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: I see.  The only other suggestion I have for making this clearer would to rename `blockSet` to `blocksIn` and `blocks` to `blocksOut` since you push things back into both of them.

Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala]: Done.  And I added more explanation to the loop for our future selves (or myself).
